articles/200_coexistence_and_migration.md

```diff
@@ -0,0 +1,247 @@
+
+### Message / service definitions
+
+In ROS 2 message / service definitions are expected to be in the subdirectory `msg` / `srv`.
+This implicitly ensures that message and service names are unique.
+There is no technical reason behind that design decision rather than simplifying the implementation.
+
+**Idea:** Allow arbitrary locations in the source space.
+
+Currently ROS 2 already supports additional feature in these files (e.g. default values) which are not available in ROS 1.
+In the future ROS 2 might support a different file format to satisfy the requested feature requests (like grouping constants in enums, optional fields, file / constant / field comments, etc.).
+If that is the case a backward compatible parser to read existing `.msg` / `.srv` files will likely be available.
+
+### Generated message / service API
+
+In order to satisfy the goal of dual-homed applications the generated code of ROS 1 and ROS 2 need to be side-by-side usable.
+
+For C++ that is already the case.
+In ROS 1 the header files are generated in namespace specific to the package name.
+In ROS 2 the header files are placed into a sub-namespace names `msg` / `srv`.
+Beside the different path / namespace the data structure also uses different types for e.g. the shared pointers (Boost vs. C++11).
+
+For Python that is not the case.
+
+**Proposal:** Change the location of the generated Python modules from using `msg` / `srv` as the subdirectory to `message` / `service`.
+At the same time updating the subdirectory name of the C++ headers would make sense to be similar.
+
+For backward compatibility the same API as for ROS 1 could be generated into a separate location.
+The user could opt-in to use these to reduce the migration effort but by sacrificing the dual-home support.
+In order to use those ROS 1 shim APIs the user would to add another path to the include directories / `PYTHONPATH`.
+
+**Idea:** Implement these additional message generators and provide a mechanism to easily use those interfaces.
+
+## Different categories of changes
+
+For the migration two different categories of changes need to be considered separately:
+
+* Changes which can be applied to ROS 1 packages to either make it compatible with ROS 2 or minimize the necessary changes.
+* Changes which are necessary to make the package work for ROS 2 but which are not compatible with ROS 1.
+
+If the second set of changes would be empty a package could be used as a ROS 1 package as well as ROS 2 packages using the same sources.
+Currently the previous sections mention a few different items which fall into the second category therefore it doesn't seem realistic to achieve compatibility using the exact same sources.
+The following sections assume that is not feasible to maintain a single source for the ROS 1 as well as the ROS 2 version of a package.
```

Comment:
"that *it* is not"

Reply:
Thanks. Fixed in c908ff0a18d10be6b5d979a37684a92c4eaf1966.